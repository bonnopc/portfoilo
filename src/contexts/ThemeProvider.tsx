import { createContext, useContext, useEffect, useState } from "react";

type Theme = "dark" | "light";

interface IThemeContext {
  theme: Theme;
  setTheme?: (theme: Theme) => void;
}

const DEFAULT_THEME: Theme = "light";

const ThemeContext = createContext<IThemeContext>({
  theme: "light",
});

export const useTheme = () => useContext(ThemeContext);

export default function ThemeProvider({ children }: any) {
  const [theme, setTheme] = useState<Theme>(DEFAULT_THEME);

  useEffect(() => {
    if (typeof document === "undefined") return;

    if (theme !== "dark" && document.body.classList.contains("dark")) {
      document.body.classList.remove("dark");
    } else if (theme === "dark" && !document.body.classList.contains("dark")) {
      document.body.classList.add("dark");
    }
  }, [theme]);

  return (
    <ThemeContext.Provider value={{ theme, setTheme }}>
      {children}
    </ThemeContext.Provider>
  );
}
