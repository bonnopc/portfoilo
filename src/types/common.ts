export type ThemeScheme = "dark" | "light";

export interface IThemeContext {
  theme: ThemeScheme;
  handleTheme?: (theme: ThemeScheme) => void;
}
