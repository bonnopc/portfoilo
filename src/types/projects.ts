import { KEY_SKILLS } from "@/config/keys";
export type ISkill = typeof KEY_SKILLS[number];

export interface IProject {
  name: string;
  responsibilities: string[];
  technologies: (ISkill | string)[];
  link?: string;
  github?: string;
  image?: string;
}
