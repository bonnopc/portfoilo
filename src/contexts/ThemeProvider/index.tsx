import { KEY_LOCAL_STORAGE_CUSTOMIZED_THEME_PREF } from "@/config/keys";
import useThemePreference from "@/hooks/useThemePreference";
import { IThemeContext, ThemeScheme } from "@/types/common";
import { createContext, useCallback, useContext, useEffect, useState } from "react";

export const DEFAULT_THEME: ThemeScheme = "light";

export const ThemeContext = createContext<IThemeContext>({
  theme: DEFAULT_THEME,
});

export const useTheme = () => useContext(ThemeContext);

export default function ThemeProvider({ children }: any) {
  const prefferedTheme: ThemeScheme = useThemePreference();
  const [theme, setTheme] = useState<ThemeScheme>(prefferedTheme);

  const handleTheme = useCallback((theme: ThemeScheme): void => {
    setTheme(theme);
    if (localStorage?.setItem) {
      localStorage.setItem(KEY_LOCAL_STORAGE_CUSTOMIZED_THEME_PREF, theme);
    }
  }, []);

  useEffect(() => {
    handleTheme(prefferedTheme);
  }, [prefferedTheme]);

  useEffect(() => {
    if (typeof document === "undefined") return;

    if (theme !== "dark" && document.body.classList.contains("dark")) {
      document.body.classList.remove("dark");
    } else if (theme === "dark" && !document.body.classList.contains("dark")) {
      document.body.classList.add("dark");
    }
  }, [theme]);

  return <ThemeContext.Provider value={{ theme, handleTheme }}>{children}</ThemeContext.Provider>;
}
