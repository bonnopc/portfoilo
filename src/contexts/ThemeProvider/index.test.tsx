import { fireEvent, render } from "@testing-library/react";
import ThemeProvider, { useTheme } from ".";
import { ThemeScheme } from "@/types/common";

const TestingComponent = () => {
  const { theme, handleTheme } = useTheme();

  const handleThemeChange = (newTheme: ThemeScheme) => {
    if (!handleTheme) return;

    handleTheme(newTheme);
  };

  return (
    <div>
      <p data-testid="theme-text" className={theme}>
        Text
      </p>
      <button onClick={() => handleThemeChange("dark")} data-testid="dark-button">
        Dark
      </button>
      <button onClick={() => handleThemeChange("light")} data-testid="light-button">
        Light
      </button>
    </div>
  );
};

describe("ThemeProvider", () => {
  it("should render", () => {
    const { container } = render(<ThemeProvider />);
    expect(container).toMatchSnapshot();
  });

  it("should render with children", () => {
    const { container } = render(<ThemeProvider>Test</ThemeProvider>);
    expect(container).toMatchSnapshot();
  });

  it("should render with children and theme", () => {
    const { container } = render(<ThemeProvider theme="dark">Test</ThemeProvider>);
    expect(container).toMatchSnapshot();
  });

  it("should render with children and theme and handleTheme", () => {
    const { container } = render(
      <ThemeProvider theme="dark" handleTheme={() => {}}>
        Test
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render with children and theme and handleTheme and className", () => {
    const { container } = render(
      <ThemeProvider theme="dark" handleTheme={() => {}} className="test">
        Test
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should change the theme state using the handleTheme function and trigger component updates", () => {
    const { getByTestId } = render(
      <ThemeProvider>
        <TestingComponent />
      </ThemeProvider>
    );
    const lightButton = getByTestId("light-button");
    const darkButton = getByTestId("dark-button");
    const themeText = getByTestId("theme-text");

    // theme to be present in the document
    expect(themeText).toBeInTheDocument();
    fireEvent.click(darkButton);
    expect(themeText?.className).toBe("dark");
    fireEvent.click(lightButton);
    expect(themeText?.className).toBe("light");
  });

  test("should add dark class to document.body when theme is dark", () => {
    const { getByTestId } = render(
      <ThemeProvider>
        <TestingComponent />
      </ThemeProvider>
    );
    const darkButton = getByTestId("dark-button");
    const themeText = getByTestId("theme-text");

    // theme to be present in the document
    expect(themeText).toBeInTheDocument();
    fireEvent.click(darkButton);
    expect(document.body.classList.contains("dark")).toBe(true);
  });
});
