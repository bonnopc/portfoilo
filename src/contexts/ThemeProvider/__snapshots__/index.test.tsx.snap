// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider should render 1`] = `<div />`;

exports[`ThemeProvider should render with children 1`] = `
<div>
  Test
</div>
`;

exports[`ThemeProvider should render with children and theme 1`] = `
<div>
  Test
</div>
`;

exports[`ThemeProvider should render with children and theme and handleTheme 1`] = `
<div>
  Test
</div>
`;

exports[`ThemeProvider should render with children and theme and handleTheme and className 1`] = `
<div>
  Test
</div>
`;
